packages/job-worker/src/playout/playout.ts

```diff
@@ -1093,51 +1102,135 @@ function timelineTriggerTimeInner(
 }
 
 export async function executeAction(context: JobContext, data: ExecuteActionProps): Promise<ExecuteActionResult> {
-	return runJobWithPlayoutCache(
+	return runJobWithPlaylistLock(
 		context,
 		// 'executeActionInner',
 		data,
-		async (cache) => {
-			const playlist = cache.Playlist.doc
+		async (playlist, lock) => {
+			if (!playlist) throw new Error(`Job playlist "${data.playlistId}" not found `)
 
 			if (!playlist.activationId) throw UserError.create(UserErrorMessage.InactiveRundown)
 			if (!playlist.currentPartInstanceId) throw UserError.create(UserErrorMessage.NoCurrentPart)
-		},
-		async (cache) => {
-			return executeActionInner(
-				context,
-				cache,
-				{
-					pieceId: data.actionDocId,
-					fromPieceType: {
-						$in: [ExpectedPackageDBType.ADLIB_ACTION, ExpectedPackageDBType.BASELINE_ADLIB_ACTION],
-					},
+
+			const initCache = await CacheForPlayoutPreInit.createPreInit(context, lock, playlist, false)
+
+			const currentPartInstance = playlist.currentPartInstanceId
+				? await context.directCollections.PartInstances.findOne(playlist.currentPartInstanceId)
+				: undefined
+			if (!currentPartInstance)
+				throw new Error(`Current PartInstance "${playlist.currentPartInstanceId}" could not be found.`)
+
+			const rundown = initCache.Rundowns.findOne(currentPartInstance.rundownId)
+			if (!rundown) throw new Error(`Current Rundown "${currentPartInstance.rundownId}" could not be found`)
+
+			const showStyle = await context.getShowStyleCompound(rundown.showStyleVariantId, rundown.showStyleBaseId)
+
+			const blueprint = await context.getShowStyleBlueprint(showStyle._id)
+
+			if (!blueprint.blueprint.executeAction && !blueprint.blueprint.executeDataStoreAction) {
+				throw UserError.create(UserErrorMessage.ActionsNotSupported)
+			}
+
+			const watchedPackagesFilter = {
+				pieceId: data.actionDocId,
+				fromPieceType: {
+					$in: [ExpectedPackageDBType.ADLIB_ACTION, ExpectedPackageDBType.BASELINE_ADLIB_ACTION],
 				},
-				async (actionContext, _rundown, _currentPartInstance, blueprint) => {
-					if (!blueprint.blueprint.executeAction) throw UserError.create(UserErrorMessage.ActionsNotSupported)
+			}
 
-					logger.info(`Executing AdlibAction "${data.actionId}": ${JSON.stringify(data.userData)}`)
+			if (blueprint.blueprint.executeDataStoreAction) {
+				const watchedPackages = await WatchedPackagesHelper.create(
```

Comment:
We appear to be doing this `WatchedPackagesHelper.create()` both here and inside executeActionInner.
It would be nice to do it once and reuse it